Edit Resolving Package Dependencies (#2300)
theforeman/foreman-documentation: guides/common/modules/con_resolving-package-dependencies.adoc

[id="Resolving_Package_Dependencies_{context}"]
= Resolving Package Dependencies

{Project} can add dependencies of packages in a Content View to the dependent repository when publishing the Content View.
To configure this, you can enable _dependency solving_.

For example, dependency solving is useful when you incrementally add a single package to a Content View version.
You might need to enable dependency solving to install that package.

> However, dependency solving is unnecessary in most situations. For example:

Reviewer: ```suggestion
However, dependency solving is unnecessary in most situations.
For example:
```

For example:

* When incrementally adding a security errata to a Content View, dependency solving can cause significant delays to Content View publication without major benefits.
* Packages from a newer erratum might have dependencies that are incompatible with packages from an older Content View version.
Incrementally adding the erratum using dependency solving might include unwanted packages.
As an alternative, consider updating the Content View.

[NOTE]
====
Dependency solving only considers packages within the repositories of the Content View.
It does not consider packages installed on clients.
ifeval::["{client-os-family}" == "Red Hat"]
For example, if a Content View includes only AppStream, dependency solving does not include dependent BaseOS content at publish time.
endif::[]

For more information, see {ContentManagementDocURL}Limitations_to_Repository_Dependency_Resolution_content-management[Limitations to Repository Dependency Resolution] in _{ContentManagementDocTitle}_.
====

Dependency solving can lead to the following problems:

> Dependency solving can significantly delay Content View publication::

Reviewer 2: ```suggestion
Significant delay in Content View publication::
```

Therefore, publish time increases with more repositories.
+

> To mitigate this problem: Use multiple Content Views with fewer repositories.
> Combine them using composite Content Views.

Reviewer: ```suggestion
To mitigate this problem, use multiple Content Views with fewer repositories and combine them into composite Content Views.
```

Author: On a scale of 1-10, how much does this bother you? :) I wrote my first version as you suggest here and elsewhere (For example, ... To mitigate, ...). Then I changed it to use colons. My reasoning was this:

* I wanted to make the descriptions easier to scan quickly: using a colon creates a label of sorts that makes it easier to scan for similar pieces of information.
* It also counts as two short sentences rather than one long sentence (=> slightly better readability score).

Reviewer: The IBM style guide which we generally use as documentation writing guide in downstream satellite recommends us to not use incomplete sentences. https://docs.engineering.redhat.com/display/CCSENABLE/Incomplete+sentences

Author: Thanks for explaining!

> {Project} does not apply content filters on dependent packages::

Reviewer 2: ```suggestion
Ignored content filters on dependent packages::
```

{Project} prioritizes resolving package dependencies over the rules in your filter.
+

> Example: If you create a filter for security purposes but enable dependency solving, {Project} can add packages that you might consider insecure.

Reviewer: ```suggestion
For example, if you create a filter for security purposes but enable dependency solving, {Project} can add packages that you might consider insecure.
```

+

> To mitigate this problem: Carefully test filtering rules to determine the required dependencies.

Reviewer: ```suggestion
To mitigate this problem, carefully test filtering rules to determine the required dependencies.
```

If dependency solving includes unwanted packages, manually identify the core basic dependencies that the extra packages and errata need.
ifeval::["{client-os-family}" == "Red Hat"]

> See also xref:Combining_exclusion_filters_with_dependency_solving[].

Reviewer: ```suggestion
For more information, see xref:Combining_exclusion_filters_with_dependency_solving[].
```

> For more information, see xref:Combining_exclusion_filters_with_dependency_solving[].

Reviewer 2: I would drop this line. And...

endif::[]

ifeval::["{client-os-family}" == "Red Hat"]
[id="Combining_exclusion_filters_with_dependency_solving"]

> .Combining exclusion filters with dependency solving

Reviewer: This should probably go into separate module.

Author: I know what you mean, that was my first instinct too. I chose to keep it all within one concept module for these reasons:

* I don't think the two modules would ever be reused separately. They should always appear together -- if you're reading about dependency solving overview (the first part of the module), you will need to follow up with reading about filtering and dependency solving.
* The first part of the module links to the second part about the filters; so they really are very closely related.
* The procedure module below (Enabling Dependency Solving) needs to point to the concept explanation to warn users about the pitfalls of the feature. I'd rather point to one concept module rather than two.

Reviewer: Well I won't object to keeping it in the same module but I would like to know what other writers think about this. @maximiliankolb @Lennonka

Reviewer 3: I agree with Aneta's view when those modules are not going to be reused and are strongly bonded to the primary procedure.

Author: @Lennonka commented elsewhere, advised to use one more example block.

So I think it's safe to resolve this thread.

====
You want to recreate {RHEL}{nbsp}8.3 using Content View filters and include selected errata from a later {RHEL}{nbsp}8 minor release.
To achieve this, you create filters to exclude most of the errata after the {RHEL}{nbsp}8.3 release date, except a few that you need.
Then, you enable dependency solving.

In this situation, dependency solving might include more packages than expected.
As a result, the client digresses from being a {RHEL}{nbsp}8.3 machine.

If you do not need the extra errata and packages, do not configure Content View filtering.
Instead, enable and use the {RHEL}{nbsp}8.3 repository on the *Content* > *Red Hat Repositories* page in the {ProjectWebUI}.
====

ifeval::["{client-os-family}" == "Red Hat"]
In some cases, excluding packages makes dependency solving impossible for DNF.
For example, if you make a {RHEL}{nbsp}8.3 repository with a few excluded packages, `dnf update` can fail.
Instead of enabling dependency solving to resolve the problem, investigate the error from `dnf` and adjust the filters to stop excluding the missing dependency.  Otherwise, dependency solving might cause the repository to digress from {RHEL} 8.3.

> endif::[]

Reviewer: Could this be rewritten to not include a specific RHEL version?

Author: I could do this:

`For example, if you make a repository based on a minor release of Red Hat Enterprise Linux with a few excluded packages, dnf update can fail. Instead of enabling dependency solving to resolve the problem, investigate the error from dnf and adjust the filters to stop excluding the missing dependency. Otherwise, dependency solving might cause the repository to digress from the minor Red Hat Enterprise Linux release.`

But it doesn't read as well as when a specific example version is used. I think it's okay to use RHEL 8.3 because the whole thing is introduced as an example, isn't it? But I don't have a strong opinion on the matter, so if you think the above version without 8.3 is better, I'll do it that way.

Reviewer: I didn't see **example** written explicitly, so if you somehow add that to what you already have that is fine. Or maybe include it in the title?

Author: There are two examples that specify RHEL 8.3.

* line 49 uses an example block; the word "example" doesn't appear in asciidoc source, only in the rendered output
* line 65 starts with "For example"

Is it possible you missed the *example* part in one of these? If so, maybe I should think about rewriting that example to make it clearer.

> Instead of enabling dependency solving to resolve the problem, investigate the error from `dnf` and adjust the filters to stop excluding the missing dependency.  Otherwise, dependency solving might cause the repository to digress from {RHEL} 8.3.

Reviewer 2: ...and use an example block for the second example too.

[id="Resolving_Package_Dependencies_{context}"]
= Resolving Package Dependencies

{Project} can add dependencies of packages in a Content View to the dependent repository when publishing the Content View.
To configure this, you can enable _dependency solving_.

For example, dependency solving is useful when you incrementally add a single package to a Content View version.
You might need to enable dependency solving to install that package.

However, dependency solving is unnecessary in most situations.
For example:

* When incrementally adding a security errata to a Content View, dependency solving can cause significant delays to Content View publication without major benefits.

> * In case of an outdated Content View version, the packages in a newer erratum might depend on packages outside that Content View.

Reviewer 4: My original sentence doesn't make sense. Let's try something like:

"A newer erratum's packages might have dependencies that are incompatible with an older content view version's packages."

Regardless of the age of the content view, there shouldn't begin to be dependencies from repositories outside of the content view (unless the user  does something silly :) )

Author: Applied

* Packages from a newer erratum might have dependencies that are incompatible with packages from an older Content View version.
Incrementally adding the erratum using dependency solving might include unwanted packages.
As an alternative, consider updating the Content View.

[NOTE]
====
Dependency solving only considers packages within the repositories of the Content View.
It does not consider packages installed on clients.
ifeval::["{client-os-family}" == "Red Hat"]
For example, if a Content View includes only AppStream, dependency solving does not include dependent BaseOS content at publish time.
endif::[]

For more information, see {ContentManagementDocURL}Limitations_to_Repository_Dependency_Resolution_content-management[Limitations to Repository Dependency Resolution] in _{ContentManagementDocTitle}_.
====

Dependency solving can lead to the following problems:

Significant delay in Content View publication::
{Project} examines every repository in a Content View for dependencies.
Therefore, publish time increases with more repositories.
+
To mitigate this problem, use multiple Content Views with fewer repositories and combine them into composite Content Views.

> Ignored content filters on dependent packages::

Reviewer 2: ```suggestion
Ignored Content View filters on dependent packages::
```
Terminological nitpick, but pls better verify with a SME.

Author: @ianballou Can you please take a look here? Can we go with Lennonka's suggestion?

Reviewer 4: "Content view filters" is better, I agree there.

{Project} prioritizes resolving package dependencies over the rules in your filter.
+
For example, if you create a filter for security purposes but enable dependency solving, {Project} can add packages that you might consider insecure.
+
To mitigate this problem, carefully test filtering rules to determine the required dependencies.
If dependency solving includes unwanted packages, manually identify the core basic dependencies that the extra packages and errata need.

ifeval::["{client-os-family}" == "Red Hat"]

> [id="Combining_exclusion_filters_with_dependency_solving"]

Reviewer: ```suggestion
[id="Combining_exclusion_filters_with_dependency_solving_{context}"]
```

.Combining exclusion filters with dependency solving
====
You want to recreate {RHEL}{nbsp}8.3 using Content View filters and include selected errata from a later {RHEL}{nbsp}8 minor release.
To achieve this, you create filters to exclude most of the errata after the {RHEL}{nbsp}8.3 release date, except a few that you need.
Then, you enable dependency solving.

In this situation, dependency solving might include more packages than expected.

> As a result, the client digresses from being a {RHEL}{nbsp}8.3 machine.

Reviewer 3: Can we further clear what the `client` means? Client OS, Hosts?

If you do not need the extra errata and packages, do not configure Content View filtering.
Instead, enable and use the {RHEL}{nbsp}8.3 repository on the *Content* > *Red Hat Repositories* page in the {ProjectWebUI}.
====
endif::[]

ifeval::["{client-os-family}" == "Red Hat"]

> [id="Excluding_packages_and_dependency_solving_with_DNF"]

Reviewer: ```suggestion
[id="Excluding_packages_and_dependency_solving_with_DNF_{context}"]
```

.Excluding packages sometimes makes dependency solving imposible for DNF
====
If you make a {RHEL}{nbsp}8.3 repository with a few excluded packages, `dnf update` can sometimes fail.

Do not enable dependency solving to resolve the problem.
Instead, investigate the error from `dnf` and adjust the filters to stop excluding the missing dependency. 

> Otherwise, dependency solving might cause the repository to digress from {RHEL} 8.3.

Reviewer 3: ```suggestion
Else, dependency solving might diverge the repository from {RHEL} 8.3.
```
We mostly use `Else`.
`Digress` is a new word for me. I am not aware if it is included in Style Guide. I would rather prefer `diverge`.
